### PR TITLE
C++20 compatibility

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Based on #6648 (CLA Signed), but heavily modified / extended:

* Implicit capture of this via [=] deprecated in C++20, and [=,this] not standard before C++20 -> now using explicit capture lists
* Implicit copy operator deprecated in gcc 9 -> add explicit '= default' definition
* std::random_shuffle deprecated in C++17 and removed in C++20 -> migrated to a replacement in RocksDB random.h API
* Add the ability to build with different std version though -DCMAKE_CXX_STANDARD=11/14/17/20 on the cmake command line
* Minimal rebuild flag of MSVC is deprecated and is forbidden with /std:c++latest (C++20)
* Added MSVC 2019 C++11 & MSVC 2019 C++20 in AppVeyor
* Added GCC 9 C++11 & GCC9 C++30 in Travis

Test plan: make check and CI

### CODE REVIEW

File: third-party/folly/folly/synchronization/Baton.h [hunk 1]
```diff
@@ -256,7 +250,9 @@ class Baton {
       const std::chrono::time_point<Clock, Duration>& deadline,
       const WaitOptions& opt) noexcept {
 
-    switch (detail::spin_pause_until(deadline, opt, [ROCKSDB_THIS_LAMBDA_CAPTURE] { return ready(); })) {
+    // Avoid implicit/explicit capture 'this' because C++ standards change
+    const auto thiz = this;
+    switch (detail::spin_pause_until(deadline, opt, [thiz] { return thiz->ready(); })) {
```
Comment (syoliver) [inline third-party/folly/folly/synchronization/Baton.h]: `[this] { return ready();` should be enough, no ?

Comment (pdillinger) [reply third-party/folly/folly/synchronization/Baton.h]: You're right. Unlike [=,this], [this] should standard (C++11 and on). Got my wires crossed

Comment (syoliver) [reply third-party/folly/folly/synchronization/Baton.h]: @pdillinger Me too... maybe removing all implicit captures can just fix the issue without ugly preprocessor define. I wanted to give a try before tomorrow but I saw your PR :)

**Comment by siying** [comment]
In the future we should add new Windows flavor in CircleCI instead.

**Comment by pdillinger** [comment]
Tab character lint issue to be fixed in #6734

**Pull request closed without merging**